### PR TITLE
[SPARK-40670][SS][PYTHON] Fix NPE in applyInPandasWithState when the input schema has "non-nullable" column(s)

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR fixes a bug which occurs NPE when the input schema of applyInPandasWithState has "non-nullable" column(s).
This PR also leaves a code comment explaining the fix. Quoting:

```
  // See processTimedOutState: we create a row which contains the actual values for grouping key,
  // but all nulls for value side by intention. This technically changes the schema of input to
  // be "nullable", hence the schema information and the internal projection of row should take
  // this into consideration. Strictly saying, it's not applied to the part of grouping key, but
  // it doesn't hurt much even if we apply the same for grouping key as well.
```

### Why are the changes needed?

There's a bug which we didn't take the non-null columns into account. This PR fixes the bug.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New UT. The new test case failed with NPE without the fix, and succeeded with the fix.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @HyukjinKwon Please take a look. Thanks in advance!

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/FlatMapGroupsInPandasWithStateExec.scala [hunk 1]
```diff
@@ -83,7 +83,17 @@ case class FlatMapGroupsInPandasWithStateExec(
   private val chainedFunc = Seq(ChainedPythonFunctions(Seq(pythonFunction)))
   private lazy val (dedupAttributes, argOffsets) = resolveArgOffsets(
     groupingAttributes ++ child.output, groupingAttributes)
-  private lazy val unsafeProj = UnsafeProjection.create(dedupAttributes, child.output)
+
+  // See processTimedOutState: we create a row which contains the actual values for grouping key,
+  // but all nulls for value side by intention. This technically changes the schema of input to
+  // be "nullable", hence the schema information and the internal projection of row should take
+  // this into consideration. Strictly saying, it's not applied to the part of grouping key, but
+  // it doesn't hurt much to we apply the same for grouping key as well.
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/FlatMapGroupsInPandasWithStateExec.scala:91]:
```suggestion
  // it doesn't hurt much even if we apply the same for grouping key as well.
```

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsInPandasWithStateSuite.scala [hunk 2]
```diff
@@ -738,4 +738,89 @@ class FlatMapGroupsInPandasWithStateSuite extends StateStoreMetricsTest {
       }
     }
   }
+
+  test("SPARK-40670: applyInPandasWithState - streaming having non-null columns") {
+    // scalastyle:off assume
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsInPandasWithStateSuite.scala:743]: Oh, let's remove this. We don't have this rule in Apache Spark.

Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsInPandasWithStateSuite.scala:743]: lol My bad.

Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsInPandasWithStateSuite.scala:743]: Looks like we have all the "style off and on" in all test cases. Maybe a separate PR?

Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/FlatMapGroupsInPandasWithStateSuite.scala:743]: sure!

**Comment by HeartSaVioR** [comment]
https://github.com/HeartSaVioR/spark/actions/runs/3193741282

^^^ There was a flaky test suite from Kafka - I reran the build separately and confirmed all builds passed.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**